services/command-router-base/src/main/java/org/eclipse/hono/commandrouter/impl/kafka/KafkaBasedMappingAndDelegatingCommandHandler.java

```diff
@@ -83,12 +100,143 @@ public void mapAndDelegateIncomingCommandMessage(final KafkaConsumerRecord<Strin
+         * commands were originally received.
+         *
+         * @param commandContext The context of the command to apply the given sendAction for.
+         * @param sendActionSupplier The Supplier for the action to send the given command to the internal Command and
+         *                           Control API endpoint provided by protocol adapters.
+         * @return A future indicating the outcome of sending the command message.
+         * @throws NullPointerException if any of the parameters is {@code null}.
+         */
+        public Future<Void> applySendCommandAction(final KafkaBasedCommandContext commandContext,
+                final Supplier<Future<Void>> sendActionSupplier) {
+            Objects.requireNonNull(commandContext);
+            Objects.requireNonNull(sendActionSupplier);
+
+            final Promise<Void> resultPromise = Promise.promise();
+            final Pair<Supplier<Future<Void>>, Promise<Void>> sendActionSupplierAndResultPromise = Pair
+                    .of(sendActionSupplier, resultPromise);
+            if (commandContext.equals(queue.peek())) {
+                sendGivenCommandAndNextInQueueIfPossible(sendActionSupplierAndResultPromise);
+            } else if (!queue.contains(commandContext)) {
+                log.warn("command won't get sent - already removed from queue [{}]", commandContext.getCommand());
+            } else {
+                // given command is not next-in-line;
+                // that means determining its target adapter instance has finished sooner (maybe because of fewer data-grid requests)
+                // compared to a command that was received earlier
+                log.debug("sending of command with offset {} delayed waiting for processing of offset {} [delayed {}]",
+                        getRecordOffset(commandContext), getRecordOffset(queue.peek()), commandContext.getCommand());
+                commandContext.getTracingSpan().log(String.format("waiting for an earlier command with offset %d to be processed first",
+                        getRecordOffset(queue.peek())));
+                commandContext.put(KEY_COMMAND_SEND_ACTION_SUPPLIER_AND_RESULT_PROMISE, sendActionSupplierAndResultPromise);
+            }
+            return resultPromise.future();
+        }
+
+        private void sendGivenCommandAndNextInQueueIfPossible(
+                final Pair<Supplier<Future<Void>>, Promise<Void>> sendActionSupplierAndResultPromise) {
+            final KafkaBasedCommandContext commandContext = queue.remove();
+            log.trace("sending [{}]", commandContext.getCommand());
+            // TODO use the supplier result future to keep track of the latest fully processed command record with regard to its partition offset
+            sendActionSupplierAndResultPromise.one().get()
+                    .onComplete(sendActionSupplierAndResultPromise.two());
+            sendNextCommandInQueueIfPossible();
```

Comment:
Wouldn't it be more appropriate to call  `sendNextCommandInQueueIfPossible()` after `sendActionSupplierAndResultPromise()` completes?

Reply:
That would make it extra sure that the command records are written to Kafka in the right order.
But I don't think it is really needed to wait for the ack from the Kafka cluster here to send the next command. I think it's enough that the `KafkaProducer.send` invocations are in the right order here, which is the case.